connectors/src/connectors/snowflake/temporal/activities.ts

```diff
@@ -1,7 +1,157 @@
+    RemoteDatabaseModel.findAll({
+      where: {
+        connectorId,
+      },
+    }),
+    RemoteSchemaModel.findAll({
+      where: {
+        connectorId,
+      },
+    }),
+    RemoteTableModel.findAll({
+      where: {
+        connectorId,
+      },
+    }),
+  ]);
+
+  const readGrantedInternalIds = new Set([
+    ...allDatabases.map((db) => db.internalId),
+    ...allSchemas.map((s) => s.internalId),
+    ...allTables
+      .filter((t) => t.permission === "selected")
+      .map((t) => t.internalId),
+  ]);
+  const tableByInternalId = Object.fromEntries(
+    allTables.map((table) => [table.internalId, table])
+  );
+
+  const parseTableInternalId = (
+    tableInternalId: string
+  ): [string, string, string] => {
+    const [dbName, schemaName, tableName] = tableInternalId.split(".");
+    if (!dbName || !schemaName || !tableName) {
+      throw new Error(`Invalid table internalId: ${tableInternalId}`);
+    }
+
+    return [dbName, schemaName, tableName];
+  };
+
+  const isTableReadGranted = (tableInternalId: string) => {
+    const [dbName, schemaName] = parseTableInternalId(tableInternalId);
+
+    const schemaInternalId = [dbName, schemaName].join(".");
+
+    return (
+      readGrantedInternalIds.has(dbName) ||
+      readGrantedInternalIds.has(schemaInternalId) ||
+      readGrantedInternalIds.has(tableInternalId)
+    );
+  };
+
+  const tablesOnSnowflake = tablesValidation.right;
+  const internalIdsOnSnowflake = new Set(
+    tablesOnSnowflake.map(
+      (t) => `${t.database_name}.${t.schema_name}.${t.name}`
+    )
+  );
+
+  for (const internalId of internalIdsOnSnowflake) {
+    if (isTableReadGranted(internalId)) {
+      let table = tableByInternalId[internalId];
+      if (!table || !table.lastUpsertedAt) {
+        if (!table) {
+          const [dbName, schemaName, tableName] =
+            parseTableInternalId(internalId);
+          table = await RemoteTableModel.create({
+            connectorId,
+            internalId,
+            permission: "inherited",
+            name: tableName,
+            schemaName,
+            databaseName: dbName,
+          });
+        }
+
+        await upsertTable({
+          dataSourceConfig: dataSourceConfigFromConnector(connector),
+          tableId: internalId,
+          tableName: internalId,
+          remoteDatabaseTableId: internalId,
+          remoteDatabaseSecretId: connector.connectionId,
+          // TODO(SNOWFLAKE): decide what to do wrt description.
+          tableDescription: "",
+          parents: [
+            `${table.databaseName}.${table.schemaName}`,
+            table.databaseName,
+          ],
+        });
+        await table.update({
+          lastUpsertedAt: new Date(),
+        });
+      }
+    }
+  }
+
+  for (const t of Object.values(tableByInternalId)) {
+    if (
+      !isTableReadGranted(t.internalId) ||
+      !internalIdsOnSnowflake.has(t.internalId)
+    ) {
+      if (t.lastUpsertedAt) {
```

Comment:
why this check ?

Reply:
if this is null then it means we didn't create the table on core yet so nothing to delete